Makefile

```diff
@@ -139,13 +144,113 @@ help-install:
+#  (1) commit all files and remove any untracked files
+#      `git status` should show nothing.
+#  (2) `make release-changelog` to print a suggested
+#      update to CHANGELOG.md. Replace "XXX" with your
+#      intended tag.
+#  (3) create an annotated tag (with -a or -s)
+#  (4) `make release-prep` which:
+#     - checks the tag
+#     - creates a virtualenv
+#     - builds and installs the sdist
+#  (5) `make release-verify` which:
+#     - runs tests
+#     - builds docker
+#  (6a) if everything succeeds: run `make release-upload`
+#       and execute the commands in order.
+#  (6b) if anything goes wrong, rollback the various steps:
+#     - delete on docker hub
+#     - delete local docker image
+#     - delete tag locally
+#     - make clean
+#
+release-check:
+	@if test -z "$(VERSION)"; then                    \
+		echo VERSION is not set.;                 \
+		echo Please create a git tag;             \
+		exit 100;                                 \
+	elif [[ "$(VERSION)" == *"dirty"* ]] ; then       \
+		echo VERSION is dirty.;                   \
+		echo Please commit all files and re-tag.; \
+		exit 101;                                 \
+	else                                              \
+		echo "===============================";   \
+		echo "Releasing version: $(VERSION)";     \
+		echo "===============================";   \
+	fi;
+
+release-env: release-env/bin/activate release-env/bin/make_shell
+
+release-env/bin/activate:
+	$(call create_venv, release-env)
+	release-env/bin/pip install -r REQUIREMENTS-CI.txt
+	touch release-env/bin/activate
+
+release-env/bin/make_shell:
+	echo '#!/bin/bash' > $@
+	echo 'source release-env/bin/activate' >> $@
+	echo 'bash "$$@"' >> $@
+	chmod a+x $@
+
+release-changelog:
+	@if test -n "$(VERSION)"; then                    \
+		echo VERSION is set to $(VERSION)         \
+		echo Create your changelog before tagging.\
+		exit 102;                                 \
+	fi;
+	@echo "##" "[XXX]" - $(shell sh -c "date +'%Y-%m-%d'")
+	@git log $(shell sh -c "git describe --tags --abbrev=0")..HEAD --pretty=format:"- %s"
+	@echo; cat CHANGELOG.md; echo "[XXX]: https://github.com/spacetx/starfish/releases/tag/XXX"
+
+release-prep: release-check release-env
+	release-env/bin/python setup.py clean
+	release-env/bin/python setup.py sdist
+	release-env/bin/pip install dist/starfish-$(VERSION).tar.gz
+
+release-verify: export SHELL=release-env/bin/make_shell
+release-verify: release-check slow release-docker
+
+release-docker: release-check
+	docker tag $(DOCKER_IMAGE) $(DOCKER_IMAGE):$(VERSION)
+	docker tag $(DOCKER_IMAGE) $(DOCKER_IMAGE):$(VERSION)-$(DOCKER_BUILD)
+
+release-upload: release-check
+	@printf '\n# Please execute the following steps\n'
+	@echo git push origin $(VERSION)
+	@echo docker push $(DOCKER_IMAGE)
+	@echo docker push $(DOCKER_IMAGE):$(VERSION)
+	@echo docker push $(DOCKER_IMAGE):$(VERSION)-$(DOCKER_BUILD)
+	@echo twine upload dist/starfish-$(VERSION).tar.gz
```

Comment:
I'm not sure exactly where, but after the entire process is complete, it probably makes sense to blow away `release-env` (or run make clean?).  Otherwise, a subsequent release will pick up the old environment.

Reply:
Yeah, `make clean` certainly does that, but of course you don't want to call it until the upload is completed. Happy to add `release-finish` until we're ready for the overarching `release` target.

Reply:
Alternatively, it can be part of the documented release steps.

It may also be worth doing something that tests for the existence of `release-env` at the start of the release process, which should be an error.

Reply:
Yup. Had that in a previous version. I'll add it back tomorrow AM and/or before merging this, whichever comes later.

Reply:
`make release-ready` now does this check.